src/FakeItEasy.Analyzer/ArgumentConstraintTypeMismatchCodeFixProvider.cs

```diff
@@ -41,27 +44,38 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                 return CompletedTask;
             }
 
-            context.RegisterCodeFix(
-                CodeAction.Create(
-                    MakeConstraintNullableCodeFixTitle,
-                    ct => this.MakeConstraintNullableAsync(context, diagnostic, ct)),
-                diagnostic);
-
-            context.RegisterCodeFix(
-                CodeAction.Create(
-                    MakeNotNullConstraintCodeFixTitle,
-                    ct => this.MakeNotNullConstraintAsync(context, diagnostic, ct)),
-                diagnostic);
+            if (diagnostic.Descriptor.Id == DiagnosticDefinitions.ArgumentConstraintNullabilityMismatch.Id)
+            {
+                context.RegisterCodeFix(
+                  CodeAction.Create(
+                      MakeConstraintNullableCodeFixTitle,
+                      ct => this.MakeConstraintNullableAsync(context, diagnostic, ct)),
+                  diagnostic);
+
+                context.RegisterCodeFix(
+                    CodeAction.Create(
+                        MakeNotNullConstraintCodeFixTitle,
+                        ct => this.MakeNotNullConstraintAsync(context, diagnostic, ct)),
+                    diagnostic);
+            }
+            else if (diagnostic.Descriptor.Id == DiagnosticDefinitions.ArgumentConstraintTypeMismatch.Id)
+            {
+                context.RegisterCodeFix(
+                    CodeAction.Create(
+                        ChangeConstraintTypeCodeFixTitle,
+                        ct => this.ChangeConstraintTypeAsync(context, diagnostic, ct)),
+                    diagnostic);

```

Comment:
I don't think we should suggest a code fix for constraints other than `A<T>._`/`A<T>.Ignored`, because changing the parameter type can cause the predicate in `A<T>.That.Matches(...)` to become invalid. This is the same reason we don't suggest a code fix (*) for argument constraint nullability mismatch on `A<T>.That.Matches(...)`.

(*) In fact we don't even report a warning in this case, because we can't provide a code fix, and it might be intentional.

Reply:
Yes, I mean non-compilable. For instance if you change the type from int to long, the predicate could work on an int but not on a long (e.g. if it calls a method that takes an int)

Reply:
Also, it's not just numeric types. You even added a test for an implicit conversion ;)

Reply:
Right! The implicit types.

Nevertheless, I still think we should report the warning and suggest the code fix for type mismatches (not nullability mismatches) in `A<T>.That.Matches(…)`. We didn't for nullability mismatches because it was possible for the complex `A<T>.That.Matches(…)` constraints to do the right thing. In the case of a type mismatch, nothing that can be put inside `Matches(…)` will ever work, so I think we should definitely report the error. And I'd rather suggest a code fix that left produced uncompiling code in this situation. There's something terribly wrong with the test, and the user should fix it. If the type of the parameter is correct, then the type of the constraint _must_ change, and if that means extensive updates to the test code, so be it.

Reply:
OK, you convinced me ;)